### ISSUE TITLE
Support Workload Identity for oauth login

### ISSUE THREAD

**Issue by MickaelFontes** [falcosecurity/falcoctl#277]
**What would you like to be added**:
Support of [workload identity](https://cloud.google.com/kubernetes-engine/docs/how-to/workload-identity) to retrieve an access token and connect to a private Artifact registry, using the `falcoctl registry auth oauth` command.

**Why is this needed**:
To take advantage of workload authentication and avoid using secrets in a GKE environment.

**Comment by MickaelFontes** [comment]
Just to add some context:
* workload identity endpoint will simply provide an access token, whether you perform a POST (with a JSON payload) or GET request at the endpoint, but you still need to have the `Metadata-Flavor: Google` header.
Example of error with `falcoctl`:
```shell
$ falcoctl registry auth oauth europe-docker.pkg.dev --token-url http://metadata.google.internal/computeMetadata/v1/instance/service-accounts/default/token -v --client-id test --client-secret secret
 ERRO  wrong client credentials, unable to retrieve token: oauth2: cannot fetch token: 403 Forbidden
       Response: Missing required header "Metadata-Flavor": "Google"
```
* The GCE metadata server is supported for auth by the [`oauth/google`](https://github.com/golang/oauth2/blob/master/google/google.go#L226) library, maybe a flag to use this library or a specific command instead of the [`oauth/clientcredentials`](https://github.com/golang/oauth2/blob/master/clientcredentials/clientcredentials.go) one could to the work.
But I don't know if the credential storage and reading solution is currently compatible with such a change.

Moreover, I'm keen on testing things or trying to add support for that, so feel free to ask me :)

**Comment by alacuku** [comment]
Hi @MickaelFontes, thanks for the detailed issue!

Just a question: does the current `falcoctl registry auth oauth` command work if you manually provide the token?

**Comment by MickaelFontes** [comment]
Hi @alacuku ! Thank you for the answer!
In fact, if I am not mistaken, to authenticate to a Google Artifact registry, you can either use an access token or a service account key, and in both cases, they are used as a regular **password** ; the username is used to specify the type of authentication (`oauth2accesstoken` for instance, see [here for full details](https://cloud.google.com/artifact-registry/docs/docker/authentication)).

So basically, I can't use the `falcoctl registry auth oauth` but I can confirm that the `falcoctl registry auth basic` works perfectly!
(with the username `oauth2accesstoken` and as a password an access token generated with the method of your choice).

I could have a cron job that regularly fetches the new access token and overwrites the `falcoctl` environment variable used for the password, but that would be a bit ugly :sweat_smile:

**Comment by loresuso** [comment]
This is really interesting; I'm having a look as well, thanks for this issue!

**Comment by max-frank** [comment]
@MickaelFontes I also have use case for using Artifact Registry so wanted to check if you still plan to/have time to work on this issue. If not I could dedicate some time to create the PR.

**Comment by MickaelFontes** [comment]
Hi @max-frank, sorry for the delay I was a bit busy.
I tested with some quick and dirty examples with the [oauth/google](https://github.com/golang/oauth2/blob/master/google/google.go#L226) library and it seemed to work well (using Workload Identity) but did not have time to make something clean.
In fact, I was wondering:
* whether it's better add a new command or just a flag for `auth basic`
* how this new auth method should be saved (simply save the registry domain and request an access token each time, or also save the access token but then comes the expiration problem...)

Following your message, I made this ([see here](https://github.com/MickaelFontes/falcoctl/commit/eaa6086923fe4f07777214d5495293adae7b89aa)) rapidly, at least `pull` works when I tested, I can add you if you want.
So yes I still plan to work on this and your help and/or guidance is very welcome!

**Comment by max-frank** [comment]
> whether it's better add a new command or just a flag for auth basic

Personally I think adding a new type of auth would be the cleaner option especially since as you mention for this one we'll need to deal with refreshing the access token every so often. 

> how this new auth method should be saved (simply save the registry domain and request an access token each time, or also save the access token but then comes the expiration problem...)

> I can add you if you want.

I am not 100% sure yet if the current code would support it nicely but we can probably be implement auth refreshing tokens using [ReuseTokenSource](https://pkg.go.dev/golang.org/x/oauth2@v0.8.0#ReuseTokenSource)  or [ReuseTokenSourceWithExpiry](https://pkg.go.dev/golang.org/x/oauth2@v0.8.0#ReuseTokenSourceWithExpiry). Then we can cache access tokens somewhere on disk to be used for later commands (and maybe make it an option to keep them in-memory only). i.e.,

1. Load cached token (might be nil)
2. Construct ReuseTokenSource with cached token 
3. Generate access token (ReuseTokenSource will do the refresh if expired)
4. Check new token against old, if they are different update cache
5. Use token for auth command  
(ideally for long running process e.g., follow step 1 and 2 should only happen once)

Also additionally it would be useful if we support the full application default credentials flow + json SA credentials i.e.,
```golang
        var tokenSource google.TokenSource
	var err error
	if gkeAuthOptions.CredentialsFile != "" {
		data, err := ioutil.ReadFile(gkeAuthOptions.CredentialsFile)
		if err != nil {
			return nil, err
		}
		tokenSource, err = google.JWTAccessTokenSourceWithScope(data, "")
		if err != nil {
			return nil, err
		}
	} else {
		tokenSource, err = google.DefaultTokenSource(ctx, "")
		if err != nil {
			return nil, err
		}
	}

	return google.ReuseTokenSource(nil, tokenSource), nil
``` 
(note i did not run/test this code)

**Comment by MickaelFontes** [comment]
Just to have a common base for discussion, the authentication methods I have in mind are listed [here](https://cloud.google.com/artifact-registry/docs/docker/authentication#before_you_begin):

* credential helpher (`gcloud` or docker dedicated helper) : if I'm not wrong, they both simply use access tokens as passwords, so nothing new but you're right that would be very nice to have them; 
* access tokens : more or less as above, with different mechanics to get or store them (docker.json, env var,etc.)
* json SA credentials : see below
 

> Also additionally it would be useful if we support the full application default credentials flow + json SA credentials i.e.

About json SA credentials, this case is already covered by the `auth basic` command right? At least that's what I think reading [that](https://cloud.google.com/artifact-registry/docs/docker/authentication#json-key), I can check if you want. 

And yes you're right, we

**Comment by loresuso** [comment]
Hello @max-frank and @MickaelFontes and really thanks for this discussion. I am not familiar with this so I have to dig deeper before expressing my opinion, give me a bit of time to do that.

In any case, there is one thing that I would like to underline:
I am super ok with using the oauth2/google library since it has support for this, but we must ensure that we fully support all the functionalities of falcoctl, namely the `artifact follow` one. With that, I mean that `falcoctl` needs to stay authenticated always, as it is running as a daemon in that case. This means that we must ensure that the logic refreshing the token is embedded in our client, like I did when implementing and supporting the client credentials OAuth2 flow. Then, it's ok to also have a specific command for that like `auth gke` or similar. 

Just wanted to point this out and hope you can take this into account while reasoning about this! In the meantime, I'm having a better look at everything, thanks again!

**Comment by max-frank** [comment]
So I've spent some hours in the falcoctl code base and it seems that either adding support for work load identity or even full docker credential helper support should be possible by just adding a new wrapper HTTPClient implementation. 

That being said integrating this new auth type (or any new auth type for that matter) nicely is a bit of a task, because of the current limits of the configuration layout/viper. So right now auth configs are saved under `registry.auth.basic` and `registry.auth.oauth` both in the from of a list of structs under those keys. Now since these are two separate lists and the registry name is only a field in the list entries there is no efficient way to simply check/get the auth config for a specific registry. Right now the flow for preparing an HTTP Client for a specific registry (implemented [here](https://github.com/falcosecurity/falcoctl/blob/main/internal/utils/registry.go#L61-L93)) is basically the following:

```mermaid
flowchart LR
    A[Get all basic auth configs]
    A --> B{reg is present}
    B -->|True| C[use for auth]
    B -->|False| D[get all oauth configs]
    D --> E{reg is present}
    E -->|True| F[use for auth]
    E -->|False| G[give up]
 ```
 
The more auth options are added the longer this flow graph gets and the more processing has to be done until the correct auth config for a registry can be found.  Now I see 3 options here

1. We simply extend the current logic and add one layer to the resolution graph. Short term this is the easiest, but its not exactly the nicest for future proofing. (Though one might argue that not a whole lot more auth options will have to be supported)
2. Change the way the registry auth configs are saved to a format like the following:
```yaml
registry:
  gcr.io:
    auth:
       workload_identity: true
   ghcr.io:
     auth:
        basic:
           username: foo
           password: bar
```

In this format it would be an O(1) operation to both get and identify the auth type to be used for a specific registry.
*BUT* this is not exactly supported by all of the underlying configuration libraries(i.e., viper) features. For loading and writing this structure to a file we could make it work by configuring a custom key delimiter when interacting with this part of the config. We have to do this because registry FQDNs will contain `.` which are used as the default delimiter. I have yet to figure out a way though to make this work with viper environment variables that are currently supported. 

3. Merge all auth config lists into one `map[registry]auth_config` on the fly to simplify at least the HTTP client creation process, i.e., while we would have to read all auth configs into memory, the check would only be a simple hash lookup.

Sorry for the long wall of text I am trying to figure out the most sensible option here 😅

**Comment by loresuso** [comment]
Thanks for the thorough analysis, that is exactly how it works now. I fully agree with you: adding another layer to that graph is not the way to go. The reason for that implementation is that we couldn't foresee other auth mechanisms at that time, but now that we have at least three different ones, we have to change it. Plus, I think we should be future-proof at this moment, since this is a special case for Workload Identity, but I guess there could be something similar for all the other cloud providers.

For point 2, totally agree that keeping the registry as key is useful for O(1) lookup, but as you said I don't see a way to implement it properly with viper. And we must be sure that env variables are supported because `falcoctl` tries to stick to [12 factor app](https://12factor.net/). Moreover, changing the layout of auths could also represent a breaking change for some users, since that format could be used in Helm charts for instance. 

So we are left with only option 3, which to me seems really fair. True that we will have everything in memory, but I don't expect users to have hundreds or thousands of auth configured, so that should be pretty manageable!

**Comment by max-frank** [comment]
Do you prefer this initial refactor be done as its own issue first before we add the workload identity part?

**Comment by loresuso** [comment]
Yes, having the refactor and then the addition of workload identity in two separate PRs could help the reviewing process, thanks

**Comment by MickaelFontes** [comment]
> Right now the flow for preparing an HTTP Client for a specific registry (implemented [here](https://github.com/falcosecurity/falcoctl/blob/main/internal/utils/registry.go#L61-L93)) is basically the following:
I just realized that when trying to connect the dots in registry.go (and realizing that and my attempt was mixing basic and oauth logics, thus making it work sometimes but clearly in a good way :sweat_smile: )

Just to be sure I understand it correctly, the idea is to kind of merge `regCred, err := ClientCredentials(reg)` and `cred, err = credentialStore.Credential(ctx, reg)` (and all potential next auth sources) into a `map[registry]auth_config` and then, if there is one,  get the right `auth_config` to create the HTTP client?

Does this change also means that the client creation `client := authn.NewClient(...)` and/or the logic inside the `NewClient()` function needs to change or not (and logic change only applies on how auth configuration is retrieved?)

**Comment by max-frank** [comment]
> Just to be sure I understand it correctly, the idea is to kind of merge regCred, err := ClientCredentials(reg) and cred, err = credentialStore.Credential(ctx, reg) (and all potential next auth sources) into a map[registry]auth_config and then, if there is one, get the right auth_config to create the HTTP client?

Ah the idea is more to change the way the auth configuration is loaded by the function so we can get rid of this chain of ifs and instead get the answer to which auth type should be used in one call. Then each auth type can provide a function to create a matching HTTP Client. Essentially the clients used and auth logic will not change just the client creation flow to make it easier to add new stuff.

**Comment by max-frank** [comment]
I spent some more time looking at what would be necessary modifications and need to revise my above answer. A proper refactor might involve a bit more than just changing the config management. I am gonna work on a PR for the refactor in the next few days.

**Comment by loresuso** [comment]
Yes @MickaelFontes, the change that I am expecting is:
- refactor to build up the map[registry]auth_config
- a change in the NewClient function to support workload identity, using the client from the oauth library (something like here  https://github.com/falcosecurity/falcoctl/blob/main/pkg/oci/authn/client.go#L49). I believe we will be able to have a client with its own TokenSource inside, so the ability to refresh tokens, if I got it right

**Comment by max-frank** [comment]
@MickaelFontes While I was working my way through the code to do some refactoring i noticed that workload identity is technically already supported to a limited degree due the way BasicAuth is implemented.

So that while auth info is kinda saved in the `falcoctl.yaml`, but this is not really used for actual authentication. Instead falcoctl will delegate the basic auth credential retrieval logic to the docker CLIs golang libraries credential store logic. This logic works basically the following way
1. Check the config file for saved credentials (see example file below)
2. If `credStore` is configured use that instead of the config file to retrieve credentials
3. If a `credentialHelper` is configured for that registry use that

Now if you configure your `~/.docker/config.json` to use either the gcloud or gcr standalone credential helper workload identity will work for `pull` and `push`. I am not sure how well this will work for `follow` though since falcoctl only retrieves the credentials once and right now does not refresh them on expire.

```json
{
        "auths": {
                "gcr.io": {},
                "foobar.io": {"username": "foo", "password": "bar"}
        },
        "credsStore": "desktop",
        "credHelpers": {
                "gcr.io": "gcloud",
        }
}
```

**Comment by MickaelFontes** [comment]
Hi @max-frank! 
I noticed this common with docker when I saw that sometimes it created a `~/.docker/config.json` file after authentication, but it did not try to exploit this common to perform auth on the push/pull commands, I'll try it thanks for the tip!

But as you underlined, it would still not be enough to support the `follow`command and is not enough to support a full classic Falco deployment using Falcoctl in the Helm values and authenticating with workload identity (my goal basically :sweat_smile: ).

I've continued to do some test to reach that and it seems that, as expected, its works :confetti_ball: ! 
(I made a dirty version with workload identity as the only token provider possible to test, it's running and I'll see if it handles well the renewal of access tokens)

However, as you said, I agree that if some rework is necessary to fully integrate credential helpers, it should be done it a separate issue/PR.
For now, I'm trying to focus on workload identity (that comes after the refactor - do not hesitate if I can be of any help) and then, it seems we have opened new horizons :sweat_smile:

**Comment by MickaelFontes** [comment]
Hi @max-frank, thanks for the update! Congrats for the work!

If I remember correctly, we wanted to go for some native support of workload identity, using same internal mechanics as in your second point. If you think is okay to go for it and it would work great with your work, I see no issue.

Do you see any implementation problem using the second option, with your refactor?

**Comment by max-frank** [comment]
Second option should be easily added to the refactor code as additional credential function as long as exceptions for being unable to retrieve a credentials are correctly resolved to empty credentials. The exception handling here could be a bit tricky for what to consider as an error and what can be considered as no credentials existing. 

So it might be worth adding the native support with the ability to enable/disable it so its only used in environments where workload identity is actually available. Should be easy enough to only call the `WithWorkloadIdentiy` (or whatever name it will be) function if some viper config key is `true`.

If we want to be even more flexible we could later add some support for defining the call order of supported credential functions.

**Comment by MickaelFontes** [comment]
Okay, I think I saw it the same way as you described: basically, `falcoctl` will try to retrieve credentials using Workload Identity only when some parameter is set to true, so we have a clear way to distinguish "no parameter is set, I use empty credentials"  VS "Workload Identity is set but does not work".
With some viper config key to distinguish those two cases.

As we described before, I don't think that native support should mean enabled by default and as you said, it should be used only when configured to use it.

One mention about the reference you mentioned @max-frank, about [credential helper simulation](https://github.com/google/go-containerregistry/tree/main/pkg/authn#emulating-cloud-provider-credential-helpers): basically it uses behind the same mechanics we first identified: [`DefaultTokenSource`](https://github.com/google/go-containerregistry/blob/v0.15.2/vendor/golang.org/x/oauth2/google/default.go#L102) using in the end [`computeTokenSource`](https://github.com/google/go-containerregistry/blob/v0.15.2/vendor/golang.org/x/oauth2/google/default.go#L167) in the case of a GKE instance.
I'll make a first version with the `ApplicationDefaultCredentials` use case, and see what it supports, but my focus remains Workload Identity support. I don't want to have too many goals in parallel and get lost in the way :sweat_smile:

**Comment by loresuso** [comment]
Hello @max-frank @MickaelFontes, everything being said here sounds about right to me.
@MickaelFontes feel free to rebase/rework https://github.com/falcosecurity/falcoctl/pull/288/files and I will be happy to review and test it! Great job everyone!

Another important thing we should take into account, is testing. 
For the oauth part, I guess we can rework and embed my PoC https://github.com/loresuso/oauth
Can we figure out together some way to test also this other part? I know it's going to be harder, but worth reasoning about it I guess :)

**Comment by MickaelFontes** [comment]
Hi @loresuso !

Thank you for your feedback! If you have some time to do a review/test it, it would be my pleasure :tada: ;)
(I'm still doing some cleanup in the printed logs and typos a bit everywhere, but for now I do not see anything to change about the core logic)

As you'll see, the new token source allows support of credentials from GCE metadata server and also from gcloud application default credentials.

About the testing part, in the end, the command added simply uses a login/password (see [here](https://github.com/falcosecurity/falcoctl/blob/9fd3a3b28371c4d0e843b16cb0f97e11f0f6b42a/pkg/oci/authn/gke.go#LL69-LL72), using an access token as a password.
The original source for credentials:
* is either a GCE metadata server simply replying with the needed access token
* or the usual `application_default_credentials.json` credential file, which is then used to get access tokens from somewhere else.
(through a usual oauth flow I imagine)

To what extent do you think we should test the whole flow :thinking:?

And last point, I think my [PR](https://github.com/falcosecurity/falcoctl/pull/288) is no longer going through the CI since I accidentally closed it while cleaning up my fork branch :sweat_smile: (`main` was not the best branch ^^)
Could you validate it another time so we can see if it passes the tests please :pray:?
